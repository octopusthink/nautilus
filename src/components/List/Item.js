import { css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React from 'react';

export const ListItem = ({ children, ...otherProps }) => {
  return <li {...otherProps}>{children}</li>;
};

ListItem.defaultProps = {
  children: undefined,
};

ListItem.propTypes = {
  /** @ignore */
  children: PropTypes.node,
};

const StyledListItem = styled(ListItem)(({ theme }) => {
  return css`
    position: relative;
    margin: 0 0 ${theme.spacing.padding.xs};
    &::before {
      position: absolute;
      right: 101%;
    }
  `;
});

export default StyledListItem;
