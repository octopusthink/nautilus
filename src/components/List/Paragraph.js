import React from 'react';

export const ListParagraph = (props) => {
  return <React.Fragment {...props} />;
};

export default ListParagraph;
