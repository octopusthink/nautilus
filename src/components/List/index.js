import { css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React from 'react';

import { bodySmall, bodyMedium, bodyLarge } from 'themes/mixins';

export const List = ({
  children,
  large,
  small,
  inverse,
  dark,
  light,
  ...otherProps
}) => {
  return <ul {...otherProps}>{children}</ul>;
};

List.defaultProps = {
  children: undefined,
  large: false,
  small: false,
  inverse: false,
  dark: false,
  light: false,
};

List.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Increase the visual prominence of the paragraph. */
  large: PropTypes.bool,
  /** Decrease the visual prominence of the paragraph. */
  small: PropTypes.bool,
  /** Inverse text colour. Used for dark backgrounds. */
  inverse: PropTypes.bool,
  /** Darken text colour. */
  dark: PropTypes.bool,
  /** Lighten text colour. */
  light: PropTypes.bool,
};

export default styled(List)(({ dark, inverse, large, light, small, theme }) => {
  return css`
      color: ${theme.colors.text.default};
      margin: 0 0 ${theme.spacing.margin.m};
      ${small && bodySmall(theme)};
      ${!small && !large && bodyMedium(theme)};
      ${large && bodyLarge(theme)};
      ${light &&
        css`
          color: ${theme.colors.text.light};
        `}

      ${dark &&
        css`
          color: ${theme.colors.text.dark};
        `}

      ${inverse &&
        css`
          color: ${theme.colors.text.inverse};

          ${light &&
            css`
              color: ${theme.colors.text.inverseLight};
            `}
          ${dark &&
            css`
              color: ${theme.colors.text.inverseDark};
            `}
        `}
    `;
});
