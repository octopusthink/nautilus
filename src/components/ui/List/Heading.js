import React from 'react';

export const ListHeading = (props) => {
  return <React.Fragment {...props} />;
};

export default ListHeading;
