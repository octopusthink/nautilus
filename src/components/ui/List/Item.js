import { css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React, { forwardRef } from 'react';

import { toUnits } from 'styles';

export const Item = forwardRef((props, ref) => {
  const { children, ...otherProps } = props;

  return (
    <li ref={ref} {...otherProps}>
      {children}
    </li>
  );
});

export const styles = (props) => {
  const { theme } = props;

  return css`
    position: relative;
    margin: 0 0 ${toUnits(theme.spacing.padding.extraSmall)};
    &::before {
      position: absolute;
      right: 101%;
    }
  `;
};

Item.defaultProps = {
  children: undefined,
};

Item.propTypes = {
  /** @ignore */
  children: PropTypes.node,
};

Item.displayName = 'List.Item';

export const { defaultProps, propTypes } = Item;

export default styled(Item)(styles);
