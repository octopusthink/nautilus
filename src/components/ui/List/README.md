A `<List>` is a collection of related elements of equal weight, presented in an itemised fashion.

Use the `<List.Item>` component for each item in your list.

```jsx
import { Heading } from '@octopusthink/nautilus';

<List ordered>
  <Heading>Things to do</Heading>
  <List.Item>Take shower. üõÄ</List.Item>
  <List.Item>Brush teeth. üòÅ</List.Item>
  <List.Item>Take over world. üåç</List.Item>
</List>
```

## Usage

Use a list to make blocks of text easier to read, and to break information into manageable chunks.

Don't use a `<List>` for:

- A collection of unrelated items. Use a `<Paragraph>` instead, or break the collection into smaller lists of related items.
- Presenting long or complex pieces of data. Use a `<Paragraph>` or a `<DataTable>` instead.
- A list of checkboxes, radio buttons, or other elements that require user interactivity. Use a `<Choice>` instead.
- A list of key/value pairs, such as a list of definitions or descriptions. Use a `<DescriptionList>` instead.

## Appearance

- Choose a bullet and numbering style and use them consistently across your site or app. Don't mix styles unless you have a good reason for doing so.
- Avoid lists of more than ten items. If you have a long list, consider breaking it up into smaller, more specific lists for more meaning and comprehension.
- Lists should use body text styling so they look identical to other body copy, especially when used in editorial copy. If a list stands alone, it can use more distinct typography.
- The space between each individual list item should be slightly bigger than the line-height inside the list item.
- Lists can be nested, but avoid too much nesting. As a general rule, nest lists to a maximum of three levels. If you need more levels, consider other ways of breaking up your content, like using an individual list for each top-level list item.
- Arrange list items in a logical way, even if they aren't sequential numbered steps. Bulleted lists can be arranged by priority, by order of importance, by size, alphabetically, numerically.

## Variations

Use a bulleted list when the order of items doesn't matter and to present content of equal status or value. Bulleted lists can be rearranged without losing any meaning.

```jsx
<List>
  <List.Item>cats</List.Item>
  <List.Item>dogs</List.Item>
  <List.Item>mice</List.Item>
</List>
```

<!--
TODO: Implement an Icon list.
Use an icon list when you want to provide additional semantic meaning‚Äîfor instance, to show dos and don'ts.

[example]
-->

Use ordered lists when item order is relevant, or to imply sequence and order. Ordered lists are commonly used when giving instructions or steps in a process.

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<List ordered>
  <Paragraph>How to file taxes:</Paragraph>
  <List.Item>Obtain wine.</List.Item>
  <List.Item>Crack open wine.</List.Item>
  <List.Item>Start drinking.</List.Item>
  <List.Item>Spread papers in an organised fashion around room.</List.Item>
  <List.Item>Reluctantly do taxes.</List.Item>
</List>
```

## Voice & Tone

- Introduce lists with a lead-in `<Paragraph>` ending in a colon (eg: "My favourite things:") or a `<Heading>` with no punctuation.
- If list items use a single word only, start each item with a lowercase letter, and do not use a full stop at the end. (eg: "cat")
- If list items use full sentences or sentence fragments (subject and verb), start each item with an uppercase letter and use a full stop at the end. (eg: "The cat ate a mouse.")
- Never use commas or semicolons at the end of list items.
- Use list items that are grammatically parallel to reflect their equal weight. For example, do not mix passive voice with active voice or declarative statements with imperative commands, and follow the same phrase structure for each item in the list.

## Accessibility

Using lists well will help improve the accessibility of your site.

List items are conveyed as a group of related elements to assistive technology users. Assistive technologies may announce the total number of items in a list, give the relevant position of each item in a list, and provide shortcuts to quickly traverse a list.

- If your list has a title or lead-in phrase, use a `<Heading>` or `<Paragraph>` component, inside the list as the first component. This will identify it as the list header using the `aria-labelledby` attribute.
- Keep list labels as short as possible to communicate your intent.
- Use a bulleted list for lists where the order isn't important, and a ordered list when it is.
- Don't use a list for stylistic or presentation needs only.
- Use semantic HTML elements (`<ol>`, `<ul>`, and `<li>`) to mark up lists. (Nautilus will handle this for you!)

Here's how you should describe a list with a `<Heading>` component:

```jsx
import { Heading } from '@octopusthink/nautilus';

<List>
  <Heading level={4}>Tasty fruit</Heading>
  <List.Item>apples üçè</List.Item>
  <List.Item>kiwis ü•ù</List.Item>
  <List.Item>mangos ü•≠</List.Item>
</List>
```

Here's how you should describe a list with a `<Paragraph>` component:

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<List ordered>
  <Paragraph>The best motorcycles, in order:</Paragraph>
  <List.Item>Honda CBR 125R</List.Item>
  <List.Item>Honda CB400 Four</List.Item>
  <List.Item>Ducati 848</List.Item>
</List>
```
