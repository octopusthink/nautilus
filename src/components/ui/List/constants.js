export const ListClassName = 'Nautilus-List';
export const ListItemClassName = 'Nautilus-ListItem';
