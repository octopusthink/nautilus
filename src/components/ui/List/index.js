import { css } from '@emotion/core';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import React, {
  Children,
  Fragment,
  cloneElement,
  useMemo,
  useState,
} from 'react';
import shortid from 'shortid';

import Heading from 'components/ui/Heading';
import Paragraph from 'components/ui/Paragraph';
import { bodyStyles } from 'styles';
import { useTheme } from 'themes';

import Item, { ComponentClassName as ItemClassName } from './Item';

export const ComponentClassName = 'Nautilus-List';

export const List = (props) => {
  const {
    children,
    className,
    dark,
    inverse,
    large,
    light,
    ordered,
    small,
    ...otherProps
  } = props;

  let ListComponent = 'ul';
  if (ordered === true) {
    ListComponent = 'ol';
  }

  const [generatedId] = useState(shortid.generate());
  const description = useMemo(() => {
    const validDescriptionComponents = Children.toArray(children).filter(
      (child) => {
        return child.type === Heading || child.type === Paragraph;
      },
    );

    if (validDescriptionComponents.length === 0) {
      return undefined;
    }

    const descriptionComponent = validDescriptionComponents[0];

    if (!descriptionComponent.props.id) {
      return cloneElement(descriptionComponent, { id: generatedId });
    }

    return descriptionComponent;
  }, [children, generatedId]);

  const items = useMemo(() => {
    return Children.toArray(children).filter((child) => {
      return child.type === Item;
    });
  }, [children]);

  const theme = useTheme();

  return (
    <Fragment>
      {description}
      <ListComponent
        aria-labelledby={description && description.props.id}
        css={css`
          ${bodyStyles({ dark, inverse, large, light, small, theme })};
          padding: 0;
          ${ordered &&
            css`
              counter-reset: list-counter;

              > .${ItemClassName} {
                list-style: none;
                counter-increment: list-counter;

                &::before {
                  content: counter(list-counter) '. ';
                }
              }
            `}
          ${!ordered &&
            css`
              > .${ItemClassName} {
                list-style: none;

                &::before {
                  content: '\\2022';
                  font-size: 0.6em;
                  line-height: 2.8;
                }
              }
            }
          `}
        `}
        className={classnames(ComponentClassName, className)}
        {...otherProps}
      >
        {items}
      </ListComponent>
    </Fragment>
  );
};

List.defaultProps = {
  children: undefined,
  className: undefined,
  large: false,
  small: false,
  inverse: false,
  dark: false,
  light: false,
  ordered: false,
};

List.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  className: PropTypes.string,
  /** Increase the visual prominence of the list. */
  large: PropTypes.bool,
  /** Decrease the visual prominence of the list. */
  small: PropTypes.bool,
  /** Inverse text colour. Used for dark backgrounds. */
  inverse: PropTypes.bool,
  /** Darken text colour. */
  dark: PropTypes.bool,
  /** Lighten text colour. */
  light: PropTypes.bool,
  /** Use numbers instead of bullets. */
  ordered: PropTypes.bool,
};

// Export Item as `List.Item`.
List.Item = Item;

export default List;
