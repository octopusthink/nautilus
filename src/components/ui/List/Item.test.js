import React from 'react';

import { render } from 'utils/testing';

import List from '.';

describe('List.Item', () => {
  it('should render an <li> tag', () => {
    const { getByTestId } = render(<List.Item data-testid="listItem">hello</List.Item>);

    expect(getByTestId('listItem').tagName).toEqual('LI');
  });

  it('should render list styles', () => {
    const { container } = render(<List.Item data-testid="listItem">hello</List.Item>);

    expect(container).toMatchSnapshot();
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = render(
      <List.Item className="foo" data-testid="listItem">
        hello
      </List.Item>,
    );

    expect(getByTestId('listItem').classList).toContain('foo');
  });
});
