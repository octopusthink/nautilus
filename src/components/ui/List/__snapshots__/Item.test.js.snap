// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List.Item should render list styles 1`] = `
.emotion-0 {
  position: relative;
  margin: 0 0 0.4rem;
}

.emotion-0::before {
  position: absolute;
  right: 101%;
}

<div>
  <li
    class="Nautilus-ListItem emotion-0"
    data-testid="listItem"
  >
    hello
  </li>
</div>
`;
